parsec/interfaces/ptg/ptg-compiler/jdf2c.c

```diff
@@ -6888,9 +6888,8 @@ static void jdf_generate_code_hook_gpu(const jdf_t *jdf,
     string_arena_free(info.sa);
 
     coutput("\n"
-            "  return parsec_%s_kernel_scheduler( es, gpu_task, dev_index );\n"
-            "}\n\n",
-            dev_lower);
+            "  return parsec_gpu_kernel_scheduler( es, gpu_task, dev_index );\n"
```

Comment:
how is this working when multiple devices are present ? I think we need to name the scheduler appropriately, to allow support for multiple devices in same time (even if we do not currently have the cmake-configury nor the code generation capability for)

Reply:
One way to revolve this would be to keep this generic function as a switchyard that would `switch(devices[dev_index]->type` and call the appropriate function based on this. That would keep the complexity in the generated code at a minimum.

Reply:
We could make the code more complicated instead of just adding the well-defined device type into the function name.  In fact, I think I would prefer a solution where the scheduler function is part of the device, so we will directly call `devices[dev_index]->scheduler(es, gpu_task, devices[dev_index])`

Reply:
Renamed in `parsec_device_kernel_scheduler` in 9cdf7db71.

When multiple accelerator devices are present, `dev_index` uniquely identifies a single device, and the code in `parsec_device_kernel_scheduler` is device-agnostic (the code uses the functions defined in the module to schedule tasklets, data movement and allocation).

Recursive and CPU devices do not call this function, as before.

Reply:
This approach limits the parsec runtime to a single enabled device per binary instance.

Reply:
Maybe the description of the PR was not clear . As part of the factorization effort, each accelerator device defines 8 functions (set_device, memcpy_async, event_query, event_record, memory_info, memory_allocate, memory_free, find_incarnation), and `parsec_device_kernel_scheduler` (and the other functions in `device_gpu.c`) use those 8 device-specific functions to call device-specific operations within a device-agnostic code.

So, `parsec_device_kernel_scheduler` in `device_gpu.c` is the generic part of the existing code that does not depend on the device type, and when we do device-specific things (e.g.  adding a CUDA event in a CUDA stream, or appending a Level Zero fence to a Level Zero command queue), we call one of the module-specific functions (gpu_module->event_record for that example).

This relies on the user / DSL putting the right thing in the `gpu_task_t` passed to `parsec_device_kernel_scheduler`: the `submit`, `stage_in`, and `stage_out` need to target the right type of device that corresponds to the `dev_index` in the call.

I believe this code should allow to have a HIP, Level Zero and CUDA device functionning simultaneously, but I may be missing something?

Reply:
it is clear. However, the only reason you need to define a `set_device` function is to be able to create a unified parsec_device_kernel_scheduler and therefore define a unique behavior across all devices. If instead you add another member to the device structure `_scheduler`, then you can in addition to having the unified behavior you describe, have specialized behavior per device (hint recursive and CPU).